File: api/routers/summary.py
"""GET /api/summary — Synthèse exécutive COMEX depuis entity_summary.jsonl."""
from __future__ import annotations

from collections import Counter

from fastapi import APIRouter

from ..loaders import load_ai_latest

router = APIRouter()


@router.get("/summary")
def get_summary():
    ai = load_ai_latest()
    entities_raw = ai.get("entity_summary", [])
    social = ai.get("social", [])
    news = ai.get("news", [])
    all_records = social + news

    # Aggregate risk_flags and opportunity_flags across all enriched records
    risk_counter: Counter = Counter()
    opp_counter: Counter = Counter()
    for r in all_records:
        for flag in r.get("risk_flags") or []:
            risk_counter[flag] += 1
        for flag in r.get("opportunity_flags") or []:
            opp_counter[flag] += 1

    top_risks = [{"flag": f, "count": c} for f, c in risk_counter.most_common(5)]
    top_opportunities = [{"flag": f, "count": c} for f, c in opp_counter.most_common(5)]

    # Sort entities by volume descending, keep top 20
    entities_sorted = sorted(entities_raw, key=lambda e: e.get("volume_items", 0), reverse=True)[:20]

    entities = [
        {
            "name": e.get("entity_name", "—"),
            "partition": e.get("source_partition", ""),
            "brand": e.get("brand_focus", "both"),
            "volume": e.get("volume_items", 0),
            "themes": (e.get("top_themes") or [])[:3],
            "risks": (e.get("top_risks") or [])[:3],
            "opportunities": (e.get("top_opportunities") or [])[:3],
            "takeaway": e.get("executive_takeaway", ""),
        }
        for e in entities_sorted
    ]

    return {
        "entities": entities,
        "top_risks": top_risks,
        "top_opportunities": top_opportunities,
    }

File: server/routes/summary.mjs
import { Router } from 'express';
import { getDb, parseJsonCol } from '../db.mjs';

const router = Router();

router.get('/summary', (req, res) => {
  const db = getDb();

  const entities = db.prepare('SELECT * FROM entity_summaries ORDER BY volume_items DESC LIMIT 20')
    .all().map(r => parseJsonCol(r, 'top_themes', 'top_risks', 'top_opportunities'));

  const social = db.prepare('SELECT risk_flags, opportunity_flags FROM social_enriched').all();
  const news = db.prepare('SELECT risk_flags, opportunity_flags FROM news_enriched').all();
  const all = [...social, ...news];

  const riskCounter = {};
  const oppCounter = {};
  for (const r of all) {
    let risks, opps;
    try { risks = typeof r.risk_flags === 'string' ? JSON.parse(r.risk_flags) : (r.risk_flags || []); } catch { risks = []; }
    try { opps = typeof r.opportunity_flags === 'string' ? JSON.parse(r.opportunity_flags) : (r.opportunity_flags || []); } catch { opps = []; }
    for (const f of risks) riskCounter[f] = (riskCounter[f] || 0) + 1;
    for (const f of opps) oppCounter[f] = (oppCounter[f] || 0) + 1;
  }

  const topRisks = Object.entries(riskCounter).sort((a, b) => b[1] - a[1]).slice(0, 5).map(([flag, count]) => ({ flag, count }));
  const topOpps = Object.entries(oppCounter).sort((a, b) => b[1] - a[1]).slice(0, 5).map(([flag, count]) => ({ flag, count }));

  res.json({
    entities: entities.map(e => ({
      name: e.entity_name || '—',
      partition: e.source_partition || '',
      brand: e.brand_focus || 'both',
      volume: e.volume_items || 0,
      themes: (e.top_themes || []).slice(0, 3),
      risks: (e.top_risks || []).slice(0, 3),
      opportunities: (e.top_opportunities || []).slice(0, 3),
      takeaway: e.executive_takeaway || '',
    })),
    top_risks: topRisks,
    top_opportunities: topOpps,
  });
});

export default router;
